keps/sig-auth/541-external-credential-providers/README.md

```diff
@@ -313,20 +337,57 @@ has completed parsing its `kubeconfig`, flags and environment variables).
 This allows the executable to perform different actions based on the current
 cluster (i.e. get a token for a particular cluster).  The `Cluster` struct is
 flexible in that it not only provides all details required to communicate with
-the cluster (hostname and TLS config), but that is also allows arbitrary
+the cluster as one would via a `kubeconfig` (i.e. everything from
+`"k8s.io/client-go/tools/clientcmd/api/v1".Cluster`), but also allows arbitrary
 per-cluster configuration to be passed to the executable via the `config` field.
 This field can contain arbitrary data that is passed to the executable without
 modification.  This allows extra user-defined data (i.e. an OAuth client ID for
 audience scoping) to be passed through the `spec.cluster` field.  The user
-configures this via the `kubeconfig`'s `clusters.[...].cluster.extensions[exec].extension`
-field.  The `exec` named extension is reserved for this purpose.
+configures this via the `kubeconfig`'s `clusters.[...].cluster.extensions[client.authentication.k8s.io/exec].extension`
+field.  The `client.authentication.k8s.io/exec` named extension is reserved for this
+purpose.
+
+To ensure that the `Cluster` struct maintains the same cluster connection
+capabilities as one gets from a `kubeconfig`, the `Cluster` fields (Go and JSON) must
+be named the same as `"k8s.io/client-go/tools/clientcmd/api/v1".Cluster`, with the
+exception of `InsecureSkipTLSVerify`, which has been left out to discourage its use, and
+`CertificateAuthority`, which has been left out since CA data will always be passed
+to the plugin as bytes.
 
 This data is passed to the executable via the `KUBERNETES_EXEC_INFO` environment
 variable in a JSON serialized object.  Note that an environment variable is used
 over passing this information via standard input because standard input is
 reserved for interactive flows between the user and executable (i.e. to prompt
 for username and password).
 
+To make it easier for a plugin to use this `KUBERNETES_EXEC_INFO` environment
+variable to connect to the referent cluster, a set of helper functions will be added
+to create a `"k8s.io/client-go/rest".Config` from the `KUBERNETES_EXEC_INFO`
+environment variable. These helper functions will live in
+`k8s.io/client-go/tools/auth/exec/exec.go` so that non-plugin developers won't pull
+in this new package and the new package can safely depend on
+`k8s.io/client-go/rest`. The helper functions are as follows.
+
+```golang
+// LoadClusterFromEnv is a helper-wrapper around LoadCluster that loads from the well-known
+// KUBERNETES_EXEC_INFO environment variable. The KUBERNETES_EXEC_INFO environment
+// variable is expected to contains a serialized client.authentication.k8s.io ExecCredential
+// object.
```

Comment:
in the implementation, account for the fact that this envvar was already used in v1alpha1 and sent a serialized ExecCredential which did not include connection info

Reply:
describe the behavior in the following scenarios (well-known error, etc?):
* the envvar is not set
* the envvar is set to an empty string
* the envvar cannot be decoded (because it is malformed)
* the envvar cannot be decoded (because it contains an unrecognized type)

or document that in `LoadCluster` and indicate this delegates to that

Reply:
might consider names like `LoadExecCredentialFromEnv` / `LoadExecCredential` instead

Reply:
ACK. I'll rev this description tonight.

Reply:
> well-known error

I am thinking here that a plugin that wants to use cluster info looks like this:

```go
func main() {
  ec, rc, err := exec.LoadExecCredentialFromEnv()
  if err != nil {
    [0]
  }
  ...
}
```

The plugin will probably do something like `fmt.Println(err); os.Exit(1);` in [0]. I can't imagine that they would want to branch on what that error return is. If the exec plugin can't get the cluster information, I would hope that it doesn't sacrifice any security properties and just quits. Seems like it makes the package API surface easier to consume if we return opaque errors from these functions. If I am wrong about this assumption, then we should be able to update this package with those error helpers in a backwards compatible manner.

> might consider names

I'll move forward with `LoadExecCredential*` to optimize for readability and speed of consensus, but I will mention that the original reason I went with `Load` is because the package name was `exec` and I had hoped that `exec.Load()` read as "load information necessary to do exec stuff".